### PR TITLE
PR2: Refactor Number Theory (1 file) + test

### PR DESCRIPTION
**Pull request by Priyansh19077**
## Summary

Chains onto PR1. Contains **1 algorithm file** — the smallest change to verify the refactoring style before the larger directories.

**`Number Theory/MillerRabinDeterministic.cpp`**
- `#include <bits/stdc++.h>` + `using namespace std;`
- Remove `static` from `_mr_pow` and `_mr_composite`
- Use `__int128` directly (no `ell` typedef)

**`tests/test_miller_rabin.cpp`** — tests known primes and composites including edge cases

## Review focus

This PR establishes the pattern used in all subsequent PRs. If the style looks good here, PRs 3–7 apply the same pattern mechanically to the remaining directories.

## Test plan

- [ ] `make -C tests run` — 1 test passes

---
_Generated by [Claude Code](https://claude.ai/code/session_017ZCiLN1ePVED473EQXZaw2)_

### CODE REVIEW

File: Number Theory/MillerRabinDeterministic.cpp [hunk 1]
```diff
@@ -1,42 +1,36 @@
...
-
-    int r = 0;
-    ll d = n - 1;
-    while ((d & 1) == 0) {
-        d >>= 1;
-        r++;
-    }
+}
 
+bool MillerRabin(__int128 n) {
```
Comment (Priyansh19077) [inline Number Theory/MillerRabinDeterministic.cpp]: Why are se having an __int128 here, it is absolutely of no use, just use long long here.

File: tests/test_miller_rabin.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    // Known composites
+    ASSERT_FALSE(MillerRabin(1));
+    ASSERT_FALSE(MillerRabin(4));
+    ASSERT_FALSE(MillerRabin(6));
+    ASSERT_FALSE(MillerRabin(9));
+    ASSERT_FALSE(MillerRabin(100));
+    ASSERT_FALSE(MillerRabin(1000000006));
+
+    // Large prime
+    ASSERT_TRUE(MillerRabin((__int128)1e18 + 9));
```
Comment (Priyansh19077) [inline tests/test_miller_rabin.cpp]: No need of having __int128 here, just 1e18 + 9 should be enough for long long.

File: Number Theory/MillerRabinDeterministic.cpp [hunk 3]
```diff
@@ -1,42 +1,36 @@
...
+    if (n < 2) return false;
+    long long d = (long long)(n - 1);
+    int s = 0;
+    while ((d & 1) == 0) { d >>= 1; s++; }
     for (int a : {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}) {
```
Comment (Priyansh19077) [inline Number Theory/MillerRabinDeterministic.cpp:31]: Can we add some comment here for better explanations? Maybe something from https://cp-algorithms.com/algebra/primality_tests.html#miller-rabin-primality-test

File: Number Theory/MillerRabinDeterministic.cpp [hunk 4]
```diff
@@ -1,42 +1,36 @@
...
-    ell res = 1; 
-    while (b > 0) {
-        if (b & 1)
-            res = (res * a) % mod; 
-        a = (a * a) % mod; 
-        b = b >> 1;
-    } 
+// O(log²N) — Deterministic Miller-Rabin primality test (correct for all N < 3.3 × 10²⁴)
+// MillerRabin(n) returns true if n is prime, false otherwise.
+// Uses __int128 for intermediate multiplication to avoid overflow.
```
Comment (Priyansh19077) [inline Number Theory/MillerRabinDeterministic.cpp]: Is this really needed? Ask yourself again once.

File: Number Theory/MillerRabinDeterministic.cpp [hunk 5]
```diff
@@ -1,42 +1,36 @@
...
-    if (x == 1 || x == n - 1)
-        return false;
+bool _mr_composite(long long n, long long a, long long d, int s) {
+    long long x = _mr_pow(a, d, n);
+    if (x == 1 || x == n - 1) return false;
     for (int r = 1; r < s; r++) {
-        x = (x * x) % n;
-        if (x == n - 1)
-            return false;
+        x = (__int128)x * x % n;
```
Comment (Priyansh19077) [inline Number Theory/MillerRabinDeterministic.cpp]: `(__int128)x * x` shouldn't this be in a bracket so that the % operator works properly. It makes it easier to read with the brackets.

**Review by Priyansh19077 — COMMENTED** [review]
Added few more comments.
